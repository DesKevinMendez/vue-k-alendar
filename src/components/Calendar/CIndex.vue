<template>
  <div class="k-alendar-container">
    <div
      v-for="calendar in monthDays"
      :key="calendar.day.toString()"
      @click="selectThisDate(calendar)"
      :class="calendar.class"
      class="date"
      :ref="(el) => (dateRefs[calendar.day] = el)"
    >
      <div class="k-alendar-span-container">
        <span class="k-alendar-text">{{ calendar.text }}</span>
        <span class="point" v-if="calendar.events.length > 0" />
      </div>
      <div class="events" v-if="calendar.events.length > 0">
        <ul>
          <KEventItem
            v-for="event in howEventsShouldRender(calendar.day, calendar.events)"
            :key="event.id"
            :event="event"
            :calendar
            @eventClicked="eventClicked"
          />
        </ul>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { useDialog } from '@/composables/useDialog'
import useEvent from '@/composables/useEvent'
import useRenderCalendar from '@/composables/useRenderCalendar'
import type { DayCalendar, KEvent, MonthDays } from '@/types'
import { ref } from 'vue'
import KEventItem from '../KEventItem.vue'

const { openEventsDetailDialog, dialogPositionToRender } = useDialog()
const { monthDays, calendarDaySelect } = useRenderCalendar()
const { eventSelected } = useEvent()
const { collision } = useDialog()
const emit = defineEmits(['eventClicked'])
const dateRefs = ref<Record<string, any>>({})

const selectThisDate = (calendar: MonthDays) => {
  const isMobile = window.innerWidth < 768

  if (isMobile) {
    const sizeOfDialog = 400
    const x = Math.floor((window.innerWidth - sizeOfDialog) / 2)
    dialogPositionToRender.value = { x, y: 16 }

    if (calendar.events.length > 0) {
      if (calendar.events.length === 1) {
        const event = calendar.events[0]

        eventSelected.value = event

        emit('eventClicked', event)
      } else {
        eventSelected.value = {
          id: 'more',
          title: '',
          start_date: '',
          description: ''
        }
      }

      calendarDaySelect.value = calendar
      openEventsDetailDialog.value = true
    }
  }
}

const eventClicked = ({
  mauseEvent,
  event,
  calendar
}: {
  mauseEvent: MouseEvent
  event: KEvent
  day: string
  calendar: DayCalendar
}) => {
  eventSelected.value = event
  calendarDaySelect.value = calendar

  if (calendar.events.length > 0) {
    let target = mauseEvent.target as HTMLElement

    if (target.tagName === 'H3') {
      target = target.parentElement as HTMLElement
    }

    dialogPositionToRender.value = collision(target)

    openEventsDetailDialog.value = true
  }

  if (event.id != 'more') {
    emit('eventClicked', event)
  }
}

const calculateEventsThatCanBeRender = (day: string) => {
  const dateDiv: HTMLDivElement | null = dateRefs.value[day]
  let heightOfEventParent = 0
  const heightOfLiElement = 32
  const paddingTopOfLiElement = 8
  const paddingOfDateDiv = 8

  let totalOfEventsThatCanRender = 0

  if (dateDiv) {
    heightOfEventParent = dateDiv.clientHeight - paddingOfDateDiv * 2

    const sizeOfSpan = (dateDiv.querySelector('span.k-alendar-text') as HTMLSpanElement)
      ?.offsetHeight
    heightOfEventParent -= sizeOfSpan || 0
    totalOfEventsThatCanRender = Math.floor(
      heightOfEventParent / (heightOfLiElement + paddingTopOfLiElement)
    )
  }

  return totalOfEventsThatCanRender
}

const howEventsShouldRender = (day: string, events: KEvent[]) => {
  const totalOfEventsThatCanRender = calculateEventsThatCanBeRender(day)

  const eventsToRender = events.slice(0, totalOfEventsThatCanRender)

  if (events.length - totalOfEventsThatCanRender > 0) {
    eventsToRender.splice(eventsToRender.length - 1, 1, {
      id: 'more',
      title: `+${events.length + 1 - totalOfEventsThatCanRender}`,
      start_date: '',
      description: ''
    })
  }

  return eventsToRender
}
</script>

<style lang="postcss" scoped>
.k-alendar-container {
  @apply grid gap-1 auto-rows-auto;
  grid-template-columns: repeat(7, minmax(0, 1fr));
}

.k-alendar-container > * {
  aspect-ratio: 1 / 1;
}

.k-alendar-span-container {
  @apply flex justify-between items-center w-full;
  .k-alendar-text {
    @apply font-bold;
  }
}

.date {
  @apply flex items-start justify-start p-2 cursor-pointer flex-col relative rounded-sm
  hover:bg-[#ebeef5] dark:hover:bg-slate-600;
  @apply border border-gray-200 dark:border-slate-600;
}

.events {
  @apply w-full hidden md:block;
  ul {
    @apply list-none space-y-2;
  }
}

.point {
  @apply w-2 h-2 bg-blue-500 rounded-full md:hidden;
}

.selected {
  @apply bg-[#ebeef5] text-inherit p-2
  dark:bg-slate-700;
}

.other-month-date {
  @apply text-gray-400;
}
</style>
