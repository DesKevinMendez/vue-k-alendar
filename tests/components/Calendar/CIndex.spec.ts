import CIndex from '@/components/Calendar/CIndex.vue'
import useRenderCalendar from '@/composables/useRenderCalendar'
import { mount, VueWrapper } from '@vue/test-utils'
import { Date2024Oct23WithoutEvents } from './../../fakes/monthDaysWithoutCEvents'
import { useDialog } from '@/composables/useDialog'
import type { KEvent } from '@/types'
import useEvent from '@/composables/useEvent'

let wrapper: VueWrapper<InstanceType<typeof CIndex>>

const events: KEvent[] = [
  {
    title: 'Event 7',
    start_date: '2024-10-01',
    description: 'Description of random event',
    color: '#677d73',
    autor: 'Kevin Méndez',
    id: 'a5ad72d6-1ab2-47a7-8870-73df8b2c6dc8'
  },
  {
    title: 'Event 8',
    start_date: '2024-10-01',
    description: 'Description of random event',
    color: '#677d73',
    autor: 'Kevin Méndez',
    id: 'a5ad72d6-1ab2-47a7-8870-73df8b2c6dc8'
  },
  {
    title: 'Event 9',
    start_date: '2024-10-01',
    description: 'Description of random event',
    color: '#677d73',
    autor: 'Kevin Méndez',
    id: 'a5ad72d6-1ab2-47a7-8870-73df8b2c6dc8'
  },
  {
    title: 'Event 10',
    start_date: '2024-10-01',
    description: 'Description of random event',
    color: '#677d73',
    autor: 'Kevin Méndez',
    id: 'a5ad72d6-1ab2-47a7-8870-73df8b2c6dc8'
  }
]

describe('CIndex', () => {
  beforeEach(() => {
    wrapper = mount(CIndex)
  })

  afterEach(() => {
    vi.clearAllMocks()
  })

  it('should render the component', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('window.innerWidth < 768 (475px)', () => {
    const originalInnnerWidth = window.innerWidth
    const innerWidth = 475

    beforeEach(() => {
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: innerWidth
      })
    })

    afterEach(() => {
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: originalInnnerWidth
      })

      vi.clearAllMocks()
    })

    const setupMonthDays = async () => {
      const { monthDays } = useRenderCalendar()
      monthDays.value = Date2024Oct23WithoutEvents
      await wrapper.vm.$nextTick()
      return monthDays
    }

    const restoreMonthDays = async (dayIndex: number) => {
      const { monthDays } = useRenderCalendar()
      monthDays.value[dayIndex].events = []
      await wrapper.vm.$nextTick()
    }

    it('date has one event', async () => {
      const monthDays = await setupMonthDays()

      const kAlendarContainer = wrapper.find('.k-alendar-container')
      const days = kAlendarContainer.findAll('.date')

      monthDays.value[1].events = events.slice(0, 1)

      await wrapper.vm.$nextTick()

      const kAlendarSpanContainer = days[1].find('.k-alendar-span-container')

      const firstSpanHasDay = kAlendarSpanContainer.find('.k-alendar-text')
      const secondSpanHasPoint = kAlendarSpanContainer.find('.point')
      expect(firstSpanHasDay.text()).toBe('1')
      expect(secondSpanHasPoint.exists()).toBeTruthy()
      monthDays.value[1].events = []
    })

    it('date has more than one event', async () => {
      const monthDays = await setupMonthDays()

      const kAlendarContainer = wrapper.find('.k-alendar-container')
      const days = kAlendarContainer.findAll('.date')

      monthDays.value[2].events = events.slice(0, 2)

      await wrapper.vm.$nextTick()

      const kAlendarSpanContainer = days[2].find('.k-alendar-span-container')

      const firstSpanHasDay = kAlendarSpanContainer.find('.k-alendar-text')
      const secondSpanHasPoint = kAlendarSpanContainer.find('.point')
      expect(firstSpanHasDay.text()).toBe('2')
      expect(secondSpanHasPoint.exists()).toBeTruthy()

      monthDays.value[1].events = []
    })

    it('do click on the event but this dont have events', async () => {
      await setupMonthDays()

      const { dialogPositionToRender } = useDialog()

      const kAlendarContainer = wrapper.find('.k-alendar-container')
      const days = kAlendarContainer.findAll('.date')

      await wrapper.vm.$nextTick()
      const kAlendarDateRow = days[1]

      expect(dialogPositionToRender.value).toStrictEqual({ x: 0, y: 0 })
      await kAlendarDateRow.trigger('click')
      expect(dialogPositionToRender.value).toStrictEqual({ x: 37, y: 16 })
    })

    it('do click on the event but this have one event', async () => {
      const event: KEvent = events[0]

      const { dialogPositionToRender, openEventsDetailDialog } = useDialog()
      const { monthDays, calendarDaySelect } = useRenderCalendar()
      const { eventSelected } = useEvent()

      monthDays.value = Date2024Oct23WithoutEvents

      const monthDayClicked = monthDays.value[1]
      await wrapper.vm.$nextTick()

      const kAlendarContainer = wrapper.find('.k-alendar-container')
      const days = kAlendarContainer.findAll('.date')

      monthDayClicked.events = events.slice(0, 1)

      await wrapper.vm.$nextTick()

      const kAlendarDateRow = days[1]

      await kAlendarDateRow.trigger('click')

      expect(eventSelected.value).toStrictEqual(event)
      expect(calendarDaySelect.value).toStrictEqual(monthDayClicked)
      expect(openEventsDetailDialog.value).toBeTruthy()
      expect(dialogPositionToRender.value).toStrictEqual({ x: 37, y: 16 })
      expect(wrapper.emitted('eventClicked')).toStrictEqual([[event]])

      monthDayClicked.events = []
    })

    it('do click on the event but this have more than one event', async () => {
      const { dialogPositionToRender, openEventsDetailDialog } = useDialog()
      const { monthDays, calendarDaySelect } = useRenderCalendar()
      const { eventSelected } = useEvent()

      monthDays.value = Date2024Oct23WithoutEvents

      const monthDayClicked = monthDays.value[1]
      await wrapper.vm.$nextTick()

      const kAlendarContainer = wrapper.find('.k-alendar-container')
      const days = kAlendarContainer.findAll('.date')

      await wrapper.vm.$nextTick()

      monthDayClicked.events = events.slice(0, 2)

      await wrapper.vm.$nextTick()

      const kAlendarDateRow = days[1]

      await kAlendarDateRow.trigger('click')

      expect(eventSelected.value).toStrictEqual({
        description: '',
        id: 'more',
        start_date: '',
        title: ''
      })
      expect(calendarDaySelect.value).toStrictEqual(monthDayClicked)
      expect(openEventsDetailDialog.value).toBeTruthy()
      expect(dialogPositionToRender.value).toStrictEqual({ x: 37, y: 16 })
      expect(wrapper.emitted('eventClicked')).toBeUndefined()

      monthDayClicked.events = []
    })
  })

  describe('window.innerWidth > 768 (1024px)', () => {
    afterEach(() => {
      vi.clearAllMocks()
    })

    const setOffsetHeightToKAlendarSpanContainer = (height: number = 24) => {
      // Setting offsetHeight to all .k-alendar-text
      wrapper.findAll('.k-alendar-span-container').forEach((span) => {
        Object.defineProperty(span.element, 'offsetHeight', {
          writable: true,
          value: height // 24 is the size of the real browser render in this size of screen (1024)
        })
      })
    }

    const setupMonthDays = async (dayIndex: number) => {
      const { monthDays } = useRenderCalendar()
      monthDays.value = Date2024Oct23WithoutEvents

      // define the height of the date row
      Object.defineProperty((wrapper.vm as any).dateRefs[`2024-10-0${dayIndex}`], 'clientHeight', {
        writable: true,
        value: 138 // 138 is the size of the real browser render in this size of screen (1024)
      })

      await wrapper.vm.$nextTick()
      return monthDays
    }

    it('date has one event', async () => {
      const rowDayCalendar = 1

      const monthDays = await setupMonthDays(rowDayCalendar)

      const kAlendarContainer = wrapper.find('.k-alendar-container')
      const days = kAlendarContainer.findAll('.date')

      monthDays.value[rowDayCalendar].events = events.slice(0, 1)

      await wrapper.vm.$nextTick()

      const kAlendarSpanContainer = days[rowDayCalendar].find('.events')
      const liEvent = kAlendarSpanContainer.findAll('li')

      expect(liEvent.length).toBe(1)
      expect(liEvent[0].text()).toBe('Event 7')

      monthDays.value[rowDayCalendar].events = []
    })

    it('date has two events', async () => {
      const rowDayCalendar = 1
      const monthDays = await setupMonthDays(1)

      const kAlendarContainer = wrapper.find('.k-alendar-container')
      const days = kAlendarContainer.findAll('.date')

      monthDays.value[rowDayCalendar].events = events.slice(0, 2)

      setOffsetHeightToKAlendarSpanContainer()

      await wrapper.vm.$nextTick()

      const kAlendarSpanContainer = days[rowDayCalendar].find('.events')
      const liEvent = kAlendarSpanContainer.findAll('li')

      expect(liEvent.length).toBe(2)
      expect(liEvent[0].text()).toBe('Event 7')
      expect(liEvent[1].text()).toBe('Event 8')

      monthDays.value[rowDayCalendar].events = []
    })

    it('date has many events and not all is filled in the row', async () => {
      const rowDayCalendar = 1
      const monthDays = await setupMonthDays(1)

      const kAlendarContainer = wrapper.find('.k-alendar-container')
      const days = kAlendarContainer.findAll('.date')

      monthDays.value[rowDayCalendar].events = events

      setOffsetHeightToKAlendarSpanContainer()

      await wrapper.vm.$nextTick()

      const kAlendarSpanContainer = days[rowDayCalendar].find('.events')
      const liEvent = kAlendarSpanContainer.findAll('li')

      expect(liEvent.length).toBe(3)
      expect(liEvent[0].text()).toBe('Event 7')
      expect(liEvent[1].text()).toBe('Event 8')
      expect(liEvent[2].text()).toBe('+2')

      monthDays.value[rowDayCalendar].events = []
    })

    it('do click on event', async () => {
      const rowDayCalendar = 1

      const { openEventsDetailDialog, dialogPositionToRender } = useDialog()
      const monthDays = await setupMonthDays(1)
      const { eventSelected } = useEvent()
      const { calendarDaySelect } = useRenderCalendar()

      dialogPositionToRender.value = { x: 100, y: 100 }

      const kAlendarContainer = wrapper.find('.k-alendar-container')
      const days = kAlendarContainer.findAll('.date')

      monthDays.value[rowDayCalendar].events = events.slice(0, 1)

      setOffsetHeightToKAlendarSpanContainer()

      await wrapper.vm.$nextTick()

      const kAlendarSpanContainer = days[rowDayCalendar].find('.events')
      const liEvent = kAlendarSpanContainer.findAll('li')

      await liEvent[0].trigger('click')
      expect(eventSelected.value).toStrictEqual(events[0])
      expect(wrapper.emitted('eventClicked')).toStrictEqual([[events[0]]])
      expect(calendarDaySelect.value).toStrictEqual(monthDays.value[rowDayCalendar])
      expect(dialogPositionToRender.value).toStrictEqual({ x: 0, y: 0 }) // collision function return this in the test
      expect(openEventsDetailDialog.value).toBeTruthy()

      monthDays.value[rowDayCalendar].events = []
    })

    it('do click on event but id of the event is "more"', async () => {
      const rowDayCalendar = 1

      const { eventSelected } = useEvent()
      const monthDays = await setupMonthDays(1)

      const kAlendarContainer = wrapper.find('.k-alendar-container')
      const days = kAlendarContainer.findAll('.date')

      monthDays.value[rowDayCalendar].events = events.slice(0, 1)

      setOffsetHeightToKAlendarSpanContainer()

      await wrapper.vm.$nextTick()

      const kAlendarSpanContainer = days[rowDayCalendar].find('.events')
      const liEvent = kAlendarSpanContainer.findAll('li')
      eventSelected.value.id = 'more'

      await liEvent[0].trigger('click')
      expect(wrapper.emitted('eventClicked')).toBeUndefined()

      monthDays.value[rowDayCalendar].events = []
    })
  })
})
