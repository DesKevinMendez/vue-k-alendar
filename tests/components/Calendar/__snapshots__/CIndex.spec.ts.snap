// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CIndex > should render the component 1`] = `"<div data-v-6548dee6="" class="k-alendar-container"></div>"`;
